File: Java/Core0/src/main/java/de/oes/core0/servicedesk0/CustomerDeparture.java
package de.oes.core0.servicedesk0;

import java.util.ArrayList;
import java.util.List;

import de.oes.core0.foundations.eVENT;
import de.oes.core0.sim.Simulator;

public class CustomerDeparture extends eVENT {

	public CustomerDeparture(Simulator sim, Long occTime, Long delay) {
		super(sim, occTime, delay);
	}

	@Override
	public List<eVENT> onEvent() {
		List<eVENT> followupEvents = new ArrayList<eVENT>();
		Simulator sim = this.getSim();
		// decrement queue length due to departure
		sim.getModel().incrementStat("queueLength", -1);
		Number queueLength = sim.getModel().getV().get("queueLength");
		// update statistics
		sim.incrementStat("departedCustomers", 1);
		if(queueLength.intValue() > 0) {
			followupEvents.add(new CustomerDeparture(sim, null, sim.getModel().getF().get("serviceTime").get().longValue()));
		}
		return followupEvents;
	}

}

File: JavaScript/Core0/Service-Desk-0/CustomerDeparture.js
class CustomerDeparture extends eVENT {
  constructor({occTime, delay}) {
    super({occTime, delay});
  }
  onEvent() {
    var followupEvents=[];
    // decrement queue length due to departure
    sim.model.v.queueLength--;
    // update statistics
    sim.stat.departedCustomers++;
    // if there are still customers waiting
    if (sim.model.v.queueLength > 0) {
      // start next service and schedule its end/departure
      followupEvents.push( new CustomerDeparture({
        delay: sim.model.f.serviceTime()
      }));
    }
    return followupEvents;
  }
}
